pkg/controller/direct/apikeys/apikeyskey_controller.go

```diff
@@ -0,0 +1,269 @@
+	MapNested(&pb.BrowserKeyRestrictions{}, &krm.KeyBrowserKeyRestrictions{}, "allowedReferrers").
+	MapNested(&pb.IosKeyRestrictions{}, &krm.KeyIosKeyRestrictions{}, "allowedBundleIds").
+	MapNested(&pb.ServerKeyRestrictions{}, &krm.KeyServerKeyRestrictions{}, "allowedIps").
+	MustBuild()
+
+type adapter struct {
+	projectID string
+	location  string
+	keyID     string
+
+	desired *krm.APIKeysKey
+	actual  *krm.APIKeysKey
+
+	gcp *api.Client
+}
+
+func (m *model) client(ctx context.Context) (*api.Client, error) {
+	var opts []option.ClientOption
+	if m.config.UserAgent != "" {
+		opts = append(opts, option.WithUserAgent(m.config.UserAgent))
+	}
+	if m.config.HTTPClient != nil {
+		opts = append(opts, option.WithHTTPClient(m.config.HTTPClient))
+	}
+	if m.config.UserProjectOverride && m.config.BillingProject != "" {
+		opts = append(opts, option.WithQuotaProject(m.config.BillingProject))
+	}
+
+	// TODO: support endpoints?
+	// if m.config.Endpoint != "" {
+	// 	opts = append(opts, option.WithEndpoint(m.config.Endpoint))
+	// }
+
+	gcpClient, err := api.NewRESTClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("building apikeys client: %w", err)
+	}
+	return gcpClient, err
+}
+
+func (m *model) AdapterForObject(ctx context.Context, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	gcp, err := m.client(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Just fetch this object?
+	obj := &krm.APIKeysKey{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	projectID := obj.Spec.ProjectRef.External
+	if projectID == "" {
+		return nil, fmt.Errorf("unable to determine project")
+	}
+
+	// TODO: Use name or request resourceID to be set on create?
+	keyID := ValueOf(obj.Spec.ResourceID)
+	if keyID == "" {
+		return nil, fmt.Errorf("unable to determine resourceID")
+	}
+
+	// This is a slightly unusual resource; the location is in the URL,
+	// but the location is always "global".
+	location := "global"
+
+	return &adapter{
+		projectID: projectID,
+		location:  location,
+		keyID:     keyID,
+		desired:   obj,
+		gcp:       gcp,
+	}, nil
+}
+
+func (a *adapter) Find(ctx context.Context) (bool, error) {
+	if a.keyID == "" {
+		return false, nil
+	}
+
+	req := &pb.GetKeyRequest{
+		Name: a.fullyQualifiedName(),
+	}
+	key, err := a.gcp.GetKey(ctx, req)
+	if err != nil {
+		if IsNotFound(err) {
+			klog.Warningf("key was not found: %v", err)
+			return false, nil
+		}
+		return false, err
+	}
+
+	u := &krm.APIKeysKey{}
+	if err := keyMapping.Map(key, u); err != nil {
+		return false, err
+	}
+	a.actual = u
+
+	return true, nil
+}
+
+func (a *adapter) Delete(ctx context.Context) (bool, error) {
+	// TODO: Delete via status selfLink?
+	req := &pb.DeleteKeyRequest{
+		Name: a.fullyQualifiedName(),
+	}
+	op, err := a.gcp.DeleteKey(ctx, req)
+	if err != nil {
+		if IsNotFound(err) {
+			return false, nil
```

Comment:
Is false actually the correct return value. If the goal of delete is for resource to not exist, then it being not found seems to indicate a success. Also if we return false on non existence then it would seems likely to trigger an endless series of retries...

Reply:
Added some comment-pointers to the interface, where it is documented as returning (true, nil) when we really deleted it, and (false, nil) when we found the object was not found.

We don't actually use that first value today, but I figured we might post different events and that I didn't want to have to go through and manually update each IsNotFound block etc.  That said, we don't need it today...